prql/src/parser.rs

```diff
@@ -93,63 +70,89 @@ fn ast_of_parse_tree(pairs: Pairs<Rule>) -> Result<Items> {
+
+                    let (name, parsed) = parsed
                         .split_first()
-                        .ok_or_else(|| anyhow!("Expected at least one item"))?
-                    {
-                        let (name, params) = name_and_params
-                            .split_first()
-                            .ok_or_else(|| anyhow!("Function requires a name."))?;
-                        // We could use `.partition` for a single pass...
-                        let positional_params = params
-                            .iter()
-                            .filter_map(|x| x.as_ident())
-                            .cloned()
-                            .collect();
-                        let named_params = params
-                            .iter()
-                            .filter_map(|x| x.as_named_arg())
-                            .cloned()
-                            .collect();
-                        Item::Function(Function {
-                            name: name
-                                .as_ident()
-                                .ok_or_else(|| {
-                                    anyhow!("Function name needs to be a word; got {name:?}")
-                                })?
-                                .to_owned(),
-                            positional_params,
-                            named_params,
-                            body: body.to_owned(),
-                        })
+                        .ok_or_else(|| anyhow!("Expected at least one item"))?;
+
+                    let name = name.clone().into_ident()?;
+
+                    let (params, body) = if let Some((Item::Expr(params), body)) = parsed.split_first(){
+                        (params, body)
                     } else {
-                        unreachable!("Expected Function, got {parsed:?}")
-                    }
+                        unreachable!("expected function params and body, got {parsed:?}")
+                    };
+
+                    let positional_params = params
+                        .iter()
+                        .filter_map(|x| x.as_ident())
+                        .cloned()
+                        .collect();
+                    let named_params = params
+                        .iter()
+                        .filter_map(|x| x.as_named_arg())
+                        .cloned()
+                        .collect();
+
+                    Item::FuncDef(FuncDef {
+                        name,
+                        positional_params,
+                        named_params,
+                        body: Box::from(body.into_only().cloned()?),
+                    })
                 }
-                Rule::function_params => Item::Expr(ast_of_parse_tree(pair.into_inner())?),
+                Rule::func_def_params => Item::Expr(ast_of_parse_tree(pair.into_inner())?),
+                Rule::func_call | Rule::func_curry  => {
```

Comment:
Changing `Terms` into the actual `FuncCall` is such an improvement; and makes other parts (e.g. all the unnesting) much simpler. Great idea!